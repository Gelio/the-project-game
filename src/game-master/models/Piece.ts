import { Point } from '../../common/Point';

export class Piece {
  public isSham: boolean;

  /**
   * When the piece is picked up, this is the position of the player
   */
  public position: Point;
  public isPickedUp: boolean;
}
