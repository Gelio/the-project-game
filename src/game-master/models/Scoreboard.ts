export class Scoreboard {
  public team1Score: number;
  public team2Score: number;
  public readonly scoreLimit: number;

  constructor(scoreLimit: number) {
    this.scoreLimit = scoreLimit;
    this.team1Score = 0;
    this.team2Score = 0;
  }
}
