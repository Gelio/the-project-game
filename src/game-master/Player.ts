import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
