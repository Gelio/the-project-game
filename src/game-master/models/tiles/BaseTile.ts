import { Player } from '../../Player';
import { Piece } from '../Piece';

export abstract class BaseTile {
  public readonly type: string;

  public x: number;
  public y: number;

  public player: Player | null = null;
  public piece: Piece | null = null;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}
