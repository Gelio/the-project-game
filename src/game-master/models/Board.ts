import { Point } from '../../common/Point';
import { BoardSize } from '../../interfaces/BoardSize';
import { Player } from '../Player';
import { Piece } from './Piece';
import { Tile } from './tiles/Tile';

export class Board {
  public readonly size: BoardSize;
  public readonly tiles: Tile[][];
  public readonly pieces: Piece[] = [];

  constructor(size: BoardSize, tiles: Tile[][]) {
    this.size = size;
    this.tiles = tiles;
  }

  public reset() {
    // TODO: implement this
  }

  public getTileAtPosition(position: Point) {
    const tileColumn = this.tiles[position.x];
    if (!tileColumn) {
      throw new Error('Invalid X coordinate');
    }

    const tile = tileColumn[position.y];
    if (!tile) {
      throw new Error('Invalid Y coordinate');
    }

    return tile;
  }

  public addPlayer(player: Player) {
    this.getTileAtPosition(player.position).player = player;
  }

  public removePlayer(point: Point) {
    this.tiles[point.x][point.y].player = null;
  }

  public movePlayer(player: Player, newPosition: Point) {
    const previousTile = this.getTileAtPosition(player.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.player !== player) {
      throw new Error('Old player position corrupted');
    }

    if (newTile.player) {
      throw new Error('Two players cannot stand on the same tile');
    }

    previousTile.player = null;
    newTile.player = player;
  }

  public addPiece(piece: Piece) {
    if (this.pieces.indexOf(piece) !== -1) {
      throw new Error('Piece already added');
    }

    const tile = this.getTileAtPosition(piece.position);
    if (tile.piece) {
      throw new Error('Piece already exists at that position');
    }

    tile.piece = piece;
    this.pieces.push(piece);
  }

  public removePiece(piece: Piece) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    this.pieces.splice(index, 1);

    if (!piece.isPickedUp) {
      const tile = this.getTileAtPosition(piece.position);
      tile.piece = null;
    }
  }

  public movePiece(piece: Piece, newPosition: Point) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    const previousTile = this.getTileAtPosition(piece.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.piece !== piece) {
      throw new Error('Old piece position corrupted');
    }

    if (newTile.piece) {
      throw new Error('Cannot move a piece on a tile which already has one');
    }

    previousTile.piece = null;
    newTile.piece = piece;
    piece.position = newPosition;
  }
}
