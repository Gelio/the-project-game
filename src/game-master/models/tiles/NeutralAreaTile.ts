import { BaseTile } from './BaseTile';

export class NeutralAreaTile extends BaseTile {}
