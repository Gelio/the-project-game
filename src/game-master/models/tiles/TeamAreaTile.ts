import { BaseTile } from './BaseTile';

export class TeamAreaTile extends BaseTile {
  public readonly type = 'TeamAreaTile';
  public hasGoal = false;
  public hasCompletedGoal = false;
}
