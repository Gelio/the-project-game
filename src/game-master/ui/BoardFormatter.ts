import { Board } from '../models/Board';
import { Piece } from '../models/Piece';
import { NeutralAreaTile } from '../models/tiles/NeutralAreaTile';
import { TeamAreaTile } from '../models/tiles/TeamAreaTile';
import { Player } from '../Player';

export class BoardFormatter {
  public displayTile(board: Board, tile: NeutralAreaTile | TeamAreaTile): string {
    if (tile.player) {
      return this.displayPlayer(tile.player);
    }

    if (tile.piece) {
      return this.displayPiece(tile.piece);
    }

    if (tile instanceof NeutralAreaTile) {
      return ' ';
    } else if (tile instanceof TeamAreaTile) {
      return this.displayTeamAreaTile(board, tile);
    }

    throw new Error('Unknown tile');
  }

  private displayPlayer(player: Player): string {
    if (player.teamId === 1) {
      return '{blue-fg}{white-bg}P{/}';
    } else {
      return '{red-fg}{white-bg}P{/}';
    }
  }

  private displayPiece(piece: Piece): string {
    if (piece.isSham) {
      return '{grey-fg}X{/}';
    }

    return '{yellow-fg}X{/}';
  }

  private displayTeamAreaTile(board: Board, tile: TeamAreaTile): string {
    const backgroundColor = tile.y < board.size.taskArea ? '{blue-bg}' : '{red-bg}';

    if (tile.hasGoal) {
      if (tile.hasCompletedGoal) {
        return `${backgroundColor}{yellow-fg}G{/}`;
      }

      return `${backgroundColor}{white-fg}G{/}`;
    }

    return `${backgroundColor} {/}`;
  }
}
