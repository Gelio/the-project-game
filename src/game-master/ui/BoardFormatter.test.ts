import { BoardFormatter } from './BoardFormatter';

import { BoardSize } from '../../interfaces/BoardSize';
import { Player } from '../Player';
import { NeutralAreaTile } from '../models/tiles/NeutralAreaTile';
import { Piece } from '../models/Piece';
import { TeamAreaTile } from '../models/tiles/TeamAreaTile';

describe('[GM] BoardFormatter', () => {
  let boardFormatter: BoardFormatter;
  let boardSize: BoardSize;

  beforeEach(() => {
    boardFormatter = new BoardFormatter();

    boardSize = {
      x: 10,
      goalArea: 10,
      taskArea: 20
    };
  });

  describe('displayTile', () => {
    describe('with player', () => {
      it('should match snapshot for player from team 1', () => {
        const player = new Player();
        player.teamId = 1;

        const tile = new NeutralAreaTile(0, 0);
        tile.player = player;

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should match snapshot for player from team 2', () => {
        const player = new Player();
        player.teamId = 2;

        const tile = new NeutralAreaTile(0, 0);
        tile.player = player;

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should return different values for players from different teams', () => {
        const player = new Player();
        player.teamId = 1;

        const tile = new NeutralAreaTile(0, 0);
        tile.player = player;

        const resultTeam1 = boardFormatter.displayTile(boardSize, tile);

        player.teamId = 2;
        const resultTeam2 = boardFormatter.displayTile(boardSize, tile);

        expect(resultTeam1).not.toBe(resultTeam2);
      });
    });

    describe('with piece', () => {
      it('should match snapshot for sham', () => {
        const piece = new Piece();
        piece.isSham = true;

        const tile = new NeutralAreaTile(0, 0);
        tile.piece = piece;

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should match snapshot for valid pieces', () => {
        const piece = new Piece();
        piece.isSham = false;

        const tile = new NeutralAreaTile(0, 0);
        tile.piece = piece;

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should return different values whether piece is a sham or not', () => {
        const piece = new Piece();
        piece.isSham = true;

        const tile = new NeutralAreaTile(0, 0);
        tile.piece = piece;

        const resultSham = boardFormatter.displayTile(boardSize, tile);

        piece.isSham = false;
        const resultValid = boardFormatter.displayTile(boardSize, tile);

        expect(resultSham).not.toBe(resultValid);
      });
    });

    describe('for team area tile', () => {
      it('should match snapshot for the first team', () => {
        const tile = new TeamAreaTile(0, 0);

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should match snapshot for the second team', () => {
        const tile = new TeamAreaTile(0, boardSize.goalArea + boardSize.taskArea);

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should match snapshot when the tile has an uncompleted goal', () => {
        const tile = new TeamAreaTile(0, 0);
        tile.hasGoal = true;
        tile.hasCompletedGoal = false;

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should match snapshot when the tile has a completed goal', () => {
        const tile = new TeamAreaTile(0, 0);
        tile.hasGoal = true;
        tile.hasCompletedGoal = true;

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toMatchSnapshot();
      });

      it('should return different values for different team areas', () => {
        const team1Tile = new TeamAreaTile(0, 0);
        const team2Tile = new TeamAreaTile(0, boardSize.goalArea + boardSize.taskArea);

        const resultTeam1 = boardFormatter.displayTile(boardSize, team1Tile);
        const resultTeam2 = boardFormatter.displayTile(boardSize, team2Tile);

        expect(resultTeam1).not.toBe(resultTeam2);
      });
    });

    describe('for neutral area tile', () => {
      it('should return a space', () => {
        const tile = new NeutralAreaTile(0, 0);

        const result = boardFormatter.displayTile(boardSize, tile);

        expect(result).toBe(' ');
      });
    });

    it('should throw an error when given an unknown tile', () => {
      expect(() => boardFormatter.displayTile(boardSize, <any>{})).toThrow();
    });
  });
});
