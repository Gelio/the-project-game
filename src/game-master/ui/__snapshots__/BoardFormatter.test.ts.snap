// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] BoardFormatter displayTile for team area tile should match snapshot for the first team 1`] = `"{blue-bg} {/}"`;

exports[`[GM] BoardFormatter displayTile for team area tile should match snapshot for the second team 1`] = `"{red-bg} {/}"`;

exports[`[GM] BoardFormatter displayTile for team area tile should match snapshot when the tile has a completed goal 1`] = `"{blue-bg}{yellow-fg}G{/}"`;

exports[`[GM] BoardFormatter displayTile for team area tile should match snapshot when the tile has an uncompleted goal 1`] = `"{blue-bg}{white-fg}G{/}"`;

exports[`[GM] BoardFormatter displayTile with piece should match snapshot for sham 1`] = `"{grey-fg}X{/}"`;

exports[`[GM] BoardFormatter displayTile with piece should match snapshot for valid pieces 1`] = `"{yellow-fg}X{/}"`;

exports[`[GM] BoardFormatter displayTile with player should match snapshot for player from team 1 1`] = `"{blue-fg}{white-bg}P{/}"`;

exports[`[GM] BoardFormatter displayTile with player should match snapshot for player from team 2 1`] = `"{red-fg}{white-bg}P{/}"`;
